File: cpp/api/src/coms/impl/zmq/RequesterZmq.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include "RequesterZmq.h"
#include "Application.h"
#include "Messages.h"
#include "JSON.h"
#include "ContextZmq.h"

namespace cameo {
namespace coms {

constexpr int SYNC_TIMEOUT = 200;

RequesterZmq::RequesterZmq() :
	m_pollingTime{100},
	m_timeout{0},
	m_contextImpl{nullptr} {

	m_canceled.store(false);
	m_timedout.store(false);
}

void RequesterZmq::setPollingTime(int value) {
	m_pollingTime = value;
}

void RequesterZmq::setTimeout(int value) {
	m_timeout = value;
}

void RequesterZmq::resetSocket() {
	if (m_requester) {
		m_requester.reset();
	}
}

void RequesterZmq::createSocket() {

	// Create a socket dealer.
	// The dealer socket can receive multiple response.
	// It also does not require to provide the identity of the recipient socket that should be done with a socket router.
	m_requester.reset(new zmq::socket_t{m_contextImpl->getContext(), zmq::socket_type::dealer});

	// Connect to the endpoint.
	m_requester->connect(m_endpoint.toString());

	// Configure the socket to not wait at close time.
	int linger {0};
	m_requester->set(zmq::sockopt::linger, linger);
}

void RequesterZmq::createAndSyncSocket(const TimeoutCounter& timeoutCounter) {

	// Memorize the timeout that can have been set before init().
	int previousTimeout = m_timeout;

	// Loop to ensure that the responder is connected to the proxy and can reply.
	m_timeout = SYNC_TIMEOUT;

	while (true) {

		// Init the socket.
		createSocket();

		// Send sync returns false if a timeout occurred.
		if (sendSync()) {
			break;
		}

		// Reset the socket in case of timeout.
		resetSocket();

		// Increase timeout.
		m_timeout += SYNC_TIMEOUT;

		// Check the global timeout.
		if (timeoutCounter.remains() == 0) {
			throw Timeout();
		}
	}

	// Reset timeout.
	m_timeout = previousTimeout;
}

bool RequesterZmq::initSocket() {

	// Reset timedout.
	m_timedout.store(false);

	if (!m_requester) {
		try {
			createAndSyncSocket(TimeoutCounter(m_timeout));
		}
		catch (const Timeout&) {
			// Timeout. As initSocket() is called in sendRequest, we prefer to not throw a timeout exception.
			m_timedout.store(true);

			// Init failed.
			return false;
		}
	}

	return true;
}

bool RequesterZmq::sendSync() {

	json::StringObject jsonRequest;
	jsonRequest.pushKey(message::TYPE);
	jsonRequest.pushValue(message::SYNC);

	// Send the request.
	sendRequest(jsonRequest.dump());

	while (true) {

		zmq::message_t message;
		if (!receiveMessage(message)) {
			return false;
		}

		if (!message.more()) {
			break;
		}
	}

	// Had a response we can exit the loop.
	return true;
}

void RequesterZmq::init(const Endpoint& endpoint, const std::string& responderIdentity, const TimeoutCounter& timeoutCounter) {

	m_endpoint = endpoint;
	m_responderIdentity = responderIdentity;

	// Get the context.
	m_contextImpl = dynamic_cast<ContextZmq *>(This::getCom().getContext());

	createAndSyncSocket(timeoutCounter);
}

RequesterZmq::~RequesterZmq() {
	resetSocket();
}

void RequesterZmq::sendRequest(const std::string& request) {

	// Init the socket if necessary.
	if (initSocket()) {

		// Start with an empty message for the dealer socket. The identity of the connected router is added by the dealer socket.
		zmq::message_t empty;
		m_requester->send(empty, zmq::send_flags::sndmore);

		zmq::message_t responderIdentityPart {m_responderIdentity.c_str(), m_responderIdentity.size()};
		m_requester->send(responderIdentityPart, zmq::send_flags::sndmore);

		m_requester->send(empty, zmq::send_flags::sndmore);

		zmq::message_t requestPart {request.c_str(), request.size()};
		m_requester->send(requestPart, zmq::send_flags::none);
	}
}

void RequesterZmq::sendRequest(const std::string& requestPart1, const std::string& requestPart2) {

	// Init the socket if necessary.
	if (initSocket()) {

		// Start with an empty message for the dealer socket.
		zmq::message_t empty;
		m_requester->send(empty, zmq::send_flags::sndmore);

		zmq::message_t responderIdentityPart {m_responderIdentity.c_str(), m_responderIdentity.size()};
		m_requester->send(responderIdentityPart, zmq::send_flags::sndmore);

		m_requester->send(empty, zmq::send_flags::sndmore);

		// Send the request in two parts.
		zmq::message_t requestPart1Part {requestPart1.c_str(), requestPart1.size()};
		m_requester->send(requestPart1Part, zmq::send_flags::sndmore);

		zmq::message_t requestPart2Part {requestPart2.c_str(), requestPart2.size()};
		m_requester->send(requestPart2Part, zmq::send_flags::none);
	}
}

void RequesterZmq::sendRequest(const std::string& requestPart1, const std::string& requestPart2, const std::string& requestPart3) {

	// Init the socket if necessary.
	if (initSocket()) {

		// Start with an empty message for the dealer socket.
		zmq::message_t empty;
		m_requester->send(empty, zmq::send_flags::sndmore);

		zmq::message_t responderIdentityPart {m_responderIdentity.c_str(), m_responderIdentity.size()};
		m_requester->send(responderIdentityPart, zmq::send_flags::sndmore);

		m_requester->send(empty, zmq::send_flags::sndmore);

		// Send the request in three parts.
		zmq::message_t requestPart1Part {requestPart1.c_str(), requestPart1.size()};
		m_requester->send(requestPart1Part, zmq::send_flags::sndmore);

		zmq::message_t requestPart2Part {requestPart2.c_str(), requestPart2.size()};
		m_requester->send(requestPart2Part, zmq::send_flags::sndmore);

		zmq::message_t requestPart3Part {requestPart3.c_str(), requestPart3.size()};
		m_requester->send(requestPart3Part, zmq::send_flags::none);
	}
}

void RequesterZmq::send(const std::string& requestData) {

	json::StringObject jsonRequest;
	jsonRequest.pushKey(message::TYPE);
	jsonRequest.pushValue(message::REQUEST);

	jsonRequest.pushKey(message::Request::APPLICATION_NAME);
	jsonRequest.pushValue(This::getName());

	jsonRequest.pushKey(message::Request::APPLICATION_ID);
	jsonRequest.pushValue(This::getId());

	jsonRequest.pushKey(message::Request::SERVER_ENDPOINT);
	jsonRequest.pushValue(This::getEndpoint().toString());

	jsonRequest.pushKey(message::Request::SERVER_PROXY_PORT);
	jsonRequest.pushValue(This::getCom().getResponderProxyPort());

	// Send the request.
	sendRequest(jsonRequest.dump(), requestData);
}

void RequesterZmq::sendTwoParts(const std::string& requestData1, const std::string& requestData2) {

	json::StringObject jsonRequest;
	jsonRequest.pushKey(message::TYPE);
	jsonRequest.pushValue(message::REQUEST);

	jsonRequest.pushKey(message::Request::APPLICATION_NAME);
	jsonRequest.pushValue(This::getName());

	jsonRequest.pushKey(message::Request::APPLICATION_ID);
	jsonRequest.pushValue(This::getId());

	jsonRequest.pushKey(message::Request::SERVER_ENDPOINT);
	jsonRequest.pushValue(This::getEndpoint().toString());

	jsonRequest.pushKey(message::Request::SERVER_PROXY_PORT);
	jsonRequest.pushValue(This::getCom().getResponderProxyPort());

	// Send the request.
	sendRequest(jsonRequest.dump(), requestData1, requestData2);
}

bool RequesterZmq::receiveMessage(zmq::message_t& message) {

	// Define the number of iterations.
	int n {0};
	if (m_pollingTime > 0) {
		n = m_timeout / m_pollingTime + 1;
	}

	// Create the poller.
	zmq::pollitem_t items[] = {
		{ *m_requester.get(), 0, ZMQ_POLLIN, 0 }
	};

	// Infinite loop if timeout is 0 or finite loop if timeout is defined.
	int i {0};
	while (i < n || m_timeout == 0) {

		// Check if the requester has been canceled.
		if (m_canceled) {
			return false;
		}

		// Poll the requester.
		zmq::poll(&items[0], 1, std::chrono::milliseconds{m_pollingTime});

		// Get a reply.
		if (items[0].revents & ZMQ_POLLIN) {
			if (!m_requester->recv(message, zmq::recv_flags::none).has_value()) {
				return false;
			}
			return true;
		}

		i++;
	}

	// Timeout.
	m_timedout.store(true);

	// Reset the socket because it cannot be reused after a timeout.
	resetSocket();

	return false;
}

std::optional<std::string> RequesterZmq::receive() {

	if (m_canceled) {
		return {};
	}


	// Receive the empty message.
	zmq::message_t empty;
	if (!receiveMessage(empty)) {
		return {};
	}

	// Receive the requester identity.
	zmq::message_t requesterIdentityPart;
	if (!receiveMessage(requesterIdentityPart)) {
		return {};
	}

	// Receive the empty message.
	if (!receiveMessage(empty)) {
		return {};
	}

	// Receive the message.
	zmq::message_t typePart;
	if (!receiveMessage(typePart)) {
		return {};
	}

	// Get the JSON type.
	json::Object jsonType;
	json::parse(jsonType, typePart);

	int type {jsonType[message::TYPE].GetInt()};

	std::optional<std::string> result;

	if (type == message::RESPONSE) {

		// Get the response part of the message.
		zmq::message_t responsePart;
		if (!receiveMessage(responsePart)) {
			return {};
		}

		result = std::string(responsePart.data<char>(), responsePart.size());
	}

	return result;
}

void RequesterZmq::cancel() {
	m_canceled.store(true);
}

bool RequesterZmq::isCanceled() {
	return m_canceled;
}

bool RequesterZmq::hasTimedout() {
	return m_timedout;
}

void RequesterZmq::terminate() {
	resetSocket();
}

}
}

File: java/api/src/main/java/eu/ill/cameo/api/coms/impl/zmq/RequesterZmq.java
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

package eu.ill.cameo.api.coms.impl.zmq;

import java.util.concurrent.atomic.AtomicBoolean;

import org.json.simple.JSONObject;

import eu.ill.cameo.api.base.This;
import eu.ill.cameo.api.base.Timeout;
import eu.ill.cameo.api.base.TimeoutCounter;
import eu.ill.cameo.api.base.impl.zmq.ContextZmq;
import eu.ill.cameo.api.coms.impl.RequesterImpl;
import eu.ill.cameo.com.Zmq;
import eu.ill.cameo.common.messages.JSON;
import eu.ill.cameo.common.messages.Messages;
import eu.ill.cameo.common.strings.Endpoint;

public class RequesterZmq implements RequesterImpl {

	private int pollingTime = 100;
	private int timeout = 0;
	
	private Zmq.Context context;
	private Zmq.Socket requester;
	private Endpoint endpoint;
	private String responderIdentity;
	
	private AtomicBoolean canceled = new AtomicBoolean(false);
	private AtomicBoolean timedout = new AtomicBoolean(false);
	
	private final static int SYNC_TIMEOUT = 200;
	
	public void setPollingTime(int value) {
		pollingTime = value;
	}

	public void setTimeout(int value) {
		timeout = value;
	}
	
	private void resetSocket() {

		// Destroy socket.
		if (requester != null) {
			context.destroySocket(requester);
			requester = null;
		}
	}
	
	private void createSocket() {
		
		// Create the socket dealer.
		// The dealer socket can receive multiple response.
		// It also does not require to provide the identity of the recipient socket that should be done with a socket router.
		requester = context.createSocket(Zmq.DEALER);
		
		requester.connect(endpoint.toString());
		
		//TODO Shall we set linger to 0?
	}
	
	private void createAndSyncSocket(TimeoutCounter timeoutCounter) {
		
		// Memorize the timeout that can have been set before init().
		int previousTimeout = timeout;
		
		timeout = SYNC_TIMEOUT;
		
		while (true) {
			
			// Init the socket.
			createSocket();
			
			// Send sync returns false if a timeout occurred.
			if (sendSync()) {
				break;
			}

			// Reset the socket in case of timeout.
			resetSocket();
			
			// Increase timeout.
			timeout += SYNC_TIMEOUT;
			
			// Check the global timeout.
			if (timeoutCounter.remains() == 0) {
				throw new Timeout();
			}
		}
		
		// Reset timeout.
		timeout = previousTimeout;
	}
	

	private boolean initSocket() {
		
		// Reset timedout.
		timedout.set(false);
		
		if (requester == null) {
			try {
				createAndSyncSocket(new TimeoutCounter(timeout));
			}
			catch (Timeout e) {
				// Timeout. As initSocket() is called in sendRequest, we prefer to not throw a timeout exception.
				timedout.set(true);
				
				// Init failed.
				return false;
			}
		}
		
		return true;
	}
	
	private boolean sendSync() {
		
		// Create the request.
		JSONObject jsonRequest = new JSONObject();
		jsonRequest.put(Messages.TYPE, Messages.SYNC);
	
		sendRequest(Messages.serialize(jsonRequest));
		if (receiveMessage() != null) {
			// Had a response we can exit the loop.
			return true;
		}
		
		return false;
	}
	
	public void init(Endpoint endpoint, String responderIdentity, TimeoutCounter timeoutCounter) {
		
		this.endpoint = endpoint;
		this.responderIdentity = responderIdentity;
		
		// Get the context.
		this.context = ((ContextZmq)This.getCom().getContext()).getContext();
		
		createAndSyncSocket(timeoutCounter);
	}
	
	private Zmq.Msg createMessage() {
		
		Zmq.Msg message = new Zmq.Msg();
		
		// Start with an empty message for the dealer socket. The identity of the connected router is added by the dealer socket.
		message.add(new byte[0]);
		message.add(responderIdentity);
		message.add(new byte[0]);
	
		return message;
	}
	
	private void sendRequest(byte[] part) {
		
		// Init the socket if necessary.
		if (initSocket()) {
			
			// Prepare and send the message.
			Zmq.Msg message = createMessage();
			message.add(part);
			
			message.send(requester);
		}
	}
	
	private void sendRequest(byte[] part1, byte[] part2) {
		
		// Init the socket if necessary.
		if (initSocket()) {
			
			// Prepare and send the message.
			Zmq.Msg message = createMessage();
			message.add(part1);
			message.add(part2);
			
			message.send(requester);
		}
	}
	
	private void sendRequest(byte[] part1, byte[] part2, byte[] part3) {
		
		// Init the socket if necessary.
		if (initSocket()) {
			
			// Prepare and send the message.
			Zmq.Msg message = createMessage();
			message.add(part1);
			message.add(part2);
			message.add(part3);
			
			message.send(requester);
		}
	}
	
	public void send(byte[] requestData) {
		
		JSONObject jsonRequest = new JSONObject();
		jsonRequest.put(Messages.TYPE, Messages.REQUEST);
		jsonRequest.put(Messages.Request.APPLICATION_NAME, This.getName());
		jsonRequest.put(Messages.Request.APPLICATION_ID, This.getId());
		jsonRequest.put(Messages.Request.SERVER_ENDPOINT, This.getEndpoint().toString());
		jsonRequest.put(Messages.Request.SERVER_PROXY_PORT, This.getCom().getResponderProxyPort());
		
		sendRequest(Messages.serialize(jsonRequest), requestData);
	}
	
	public void send(String request) {
		send(Messages.serialize(request));
	}
	
	public void sendTwoParts(byte[] requestData1, byte[] requestData2) {
		
		JSONObject jsonRequest = new JSONObject();
		jsonRequest.put(Messages.TYPE, Messages.REQUEST);
		jsonRequest.put(Messages.Request.APPLICATION_NAME, This.getName());
		jsonRequest.put(Messages.Request.APPLICATION_ID, This.getId());
		jsonRequest.put(Messages.Request.SERVER_ENDPOINT, This.getEndpoint().toString());
		jsonRequest.put(Messages.Request.SERVER_PROXY_PORT, This.getCom().getResponderProxyPort());
		
		sendRequest(Messages.serialize(jsonRequest), requestData1, requestData2);
	}

	private Zmq.Msg receiveMessage() {
		
		// Define the number of iterations.
		int n = 0;
		if (pollingTime > 0) {
			n = timeout / pollingTime + 1;
		}

		// Create the poller.
		Zmq.Poller poller = this.context.createPoller(1);
		poller.register(requester);
		
		// Infinite loop if timeout is 0 or finite loop if timeout is defined.
		int i = 0;
		while (i < n || timeout == 0) {

			// Check if the requester has been canceled.
			if (canceled.get()) {
				return null;
			}

			// Poll the requester.
			poller.poll(pollingTime);
			if (poller.pollin(0)) {
				return Zmq.Msg.recvMsg(requester);
			}

			i++;
		}

		// Timeout occurred.
		timedout.set(true);

		// Reset the socket because it cannot be reused after a timeout.
		resetSocket();
		
		return null;
	}
	
	public byte[] receive() {
		
		Zmq.Msg message = null;
		
		try {
			message = receiveMessage();

			if (message == null) {
				return null;
			}

			// Get the data.
			byte[][] data = message.getAllData();
			
			// Get the JSON request object.
			JSONObject request = This.getCom().parse(data[3]);
			
			// Get the type.
			long type = JSON.getLong(request, Messages.TYPE);
						
			if (type == Messages.RESPONSE) {
				return data[4];
			}
			else {
				return null;
			}
		}
		finally {
			if (message != null) {
				message.destroy();
			}
		}
	}

	public String receiveString() {
		
		byte[] response = receive();
		
		if (response != null) {
			return Messages.parseString(response);
		}
		
		return null;
	}
	
	public void cancel() {
		canceled.set(true);		
	}

	public boolean isCanceled() {
		return canceled.get();
	}

	@Override
	public boolean hasTimedout() {
		return timedout.get();
	}
	
	public void terminate() {
		resetSocket();
	}
	
}